lib/pymedphys/_experimental/streamlit/apps/dicom_explorer.py

```diff
@@ -0,0 +1,62 @@
+# Copyright (C) 2021 Cancer Care Associates
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import textwrap
+
+from pymedphys._imports import pydicom
+from pymedphys._imports import streamlit as st
+
+from pymedphys._streamlit import categories
+
+CATEGORY = categories.PLANNING
+TITLE = "DICOM Explorer"
+
+
+def main():
+    left_column, right_column = st.beta_columns(2)
+
+    with left_column:
+        st.write("## Upload DICOM file")
+        dicom_bytes = st.file_uploader("DICOM file")
+
+    if not dicom_bytes:
+        raise st.stop()
+
+    try:
+        dicom_header = pydicom.dcmread(dicom_bytes, force=True, stop_before_pixels=True)
+    except Exception as e:
+        st.write(e)
+        st.stop()
+        raise
+
+    with right_column:
+        st.write(
+            f"""
+
+            ## File overview
+
+            * Patient ID: `{dicom_header.PatientID}`
+            * Patient Name: `{dicom_header.PatientName}`
+            """
+        )
+
+    st.markdown(
+        textwrap.dedent(
+            """\
+            ```
+            {}
+            ```
+            """
+        ).format(dicom_header.__repr__())
```

Comment:
@SimonBiggs. Just for my understanding, is `__repr()__` needed here? Did you just include it for explicitness?

Reply:
Without calling `__repr__` internally `__str__` would have been called by default.

I probably should have called it as `repr(dicom_header)` instead of using the dunder call.

...In the past on pydicom I thought `str()` and `repr()` gave a different response. But it seems now that they're essentially the same.

Some objects do have different `repr` and `str`, and strictly `repr` as opposed to string is what is wanted here. When they are different it's because `repr` is designed to be a human friendly representation of the object, and `str` is supposed to be "what the value of the object would be if it was converted to string".

Given here, I want a "representation of the object that makes sense for humans" `repr` is the way to go here.

Here is an example of the difference with an object we use a lot:

```ipython
C:\Users\sbiggs\git\pymedphys [dicom-viewer ≡ +0 ~1 -0 !]> ipython
Python 3.7.9 (tags/v3.7.9:13c94747c7, Aug 17 2020, 18:58:18) [MSC v.1900 64 bit (AMD64)]
Type 'copyright', 'credits' or 'license' for more information
IPython 7.19.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import numpy as np

In [2]: x = np.array([1, 2, 3])

In [3]: str(x)
Out[3]: '[1 2 3]'

In [4]: repr(x)
Out[4]: 'array([1, 2, 3])'
```

Reply:
Is it possible that you have `str` and `repr` the wrong way around? My reading of the difference between `str` and `repr` is along the lines of:

- `repr` is meant to produce an"official" string representation of an object, and if possible, should look like a valid Python expression that could be used to recreate an object with the same value. It's goal is to be unambiguous and major use case is for debugging
- `str` is meant to produce a readable representation of the object. Its goal is readability and major use case is printing.

These SO links seems to corroborate this:
- https://stackoverflow.com/a/2626364/4665458
- https://stackoverflow.com/a/3691117/4665458
- https://stackoverflow.com/a/18393727/4665458

Seeing as the purpose of this app seems to be to produce a human readable version of the DICOM dataset, it seems to me that `str` might actually be more appropriate than `repr`, at least idiomatically.

Importantly, I don't really mind at all. Just learning! :smile:

Reply:
Ahh, you're right! I do have it the wrong way round!